src/base/net/downloadhandler.cpp

```diff
@@ -174,29 +178,29 @@ void Net::DownloadHandler::handleRedirection(QUrl newUrl)
             emit downloadFailed(m_downloadRequest.url(), tr("Unexpected redirect to magnet URI."));
 
         this->deleteLater();
+        return;
     }
-    else {
-        DownloadHandler *redirected = m_manager->download(DownloadRequest(m_downloadRequest).url(newUrlString));
-        connect(redirected, &DownloadHandler::destroyed, this, &DownloadHandler::deleteLater);
-        connect(redirected, &DownloadHandler::downloadFailed, this, [this](const QString &, const QString &reason)
-        {
-            emit downloadFailed(url(), reason);
-        });
-        connect(redirected, &DownloadHandler::redirectedToMagnet, this, [this](const QString &, const QString &magnetUri)
-        {
-            emit redirectedToMagnet(url(), magnetUri);
-        });
-        connect(redirected, static_cast<void (DownloadHandler::*)(const QString &, const QString &)>(&DownloadHandler::downloadFinished)
-                , this, [this](const QString &, const QString &fileName)
-        {
-            emit downloadFinished(url(), fileName);
-        });
-        connect(redirected, static_cast<void (DownloadHandler::*)(const QString &, const QByteArray &)>(&DownloadHandler::downloadFinished)
-                , this, [this](const QString &, const QByteArray &data)
-        {
-            emit downloadFinished(url(), data);
-        });
-    }
+
+    const DownloadHandler *redirected = m_manager->download(DownloadRequest(m_downloadRequest).url(newUrlString).incrementRedirectionCounter());
+    connect(redirected, &DownloadHandler::destroyed, this, &DownloadHandler::deleteLater);
+    connect(redirected, &DownloadHandler::downloadFailed, this, [this](const QString &, const QString &reason)
+    {
+        emit downloadFailed(url(), reason);
+    }, Qt::QueuedConnection);  // to avoid nested calls
```

Comment:
Is it really necessary? Please describe. You did limit it to 20.

Reply:
I think yes, recursion building up on stack is never a good thing.

Reply:
>recursion building up on stack is never a good thing.

Isn't any recursion there at the expense of the stack? Or are you against recursion as such?
AFAIK, recursion is dangerous only when it is uncontrollable (i.e. when there are no restrictions on its depth). "Event loop" can also be fraught with some dangers.
Are you sure you checked it out well enough?

Reply:
>Or are you against recursion as such?

It should be avoided whenever possible.

>when there are no restrictions on its depth

And how do you estimate how deep is too much? Even if there is an estimation it will be dependent of the exact usage context (how many stack space left).

Reply:
@Chocobo1: 

>Even if there is an estimation it will be dependent of the exact usage context 

Why not make the exact number of recursions an Advanced Option then? 

Default to 20, but give users the choice to set it from 0 to 128, or something similar to that.

Reply:
>It should be avoided whenever possible.

Another witch hunt?
Recursion is quite good and convenient tool. It has its pros and cons, so we just need to use it correctly (as well as other tools).
The only mistake that was here was the uncontrollable depth.
>Even if there is an estimation it will be dependent of the exact usage context (how many stack space left).

It's not the problem of recursion itself. Compare trivial recursive function with couple of local variables that is limited to 10-20 subcalls with other non-recursive function having many locals.

P.S. I don't insist you change it here. I just want you to have no illusions about your solution, but to test it well. As I said above, "Event loop" can also be fraught with some dangers.

Reply:
>Why not make the exact number of recursions an Advanced Option then?

The recursion is not a goal, but how many redirections, AFAIK web browsers hard coded to 20 too, so I don't think it need to be tune-able.

>Another witch hunt?

?? What is the previous :)
Funny that you declare something a witch hunt based on your opinion...

>It's not the problem of recursion itself.

Recursion are widely used in functional programming languages, it makes perfectly sense there, but they also suffer from space leak when not careful, and so tail-call optimization should be employed whenever possible. (ha, you probably going to freak out on my statement).

>I just want you to have no illusions about your solution, but to test it well.

Back on track, using QueuedConnection here is one easy solution, and I did not find any downside so far, if there are other better alternatives I'm listening.

>As I said above, "Event loop" can also be fraught with some dangers.

Now the downloadFailed is executed sequentially and queued on event loop, if you have some concerns please speak out.

Reply:
>Funny that you declare something a witch hunt based on your opinion...

That's just my personal opinion. When I see some slogans like "Recursion is evil", "Non-const variables is evil" I consider it as some kind of "wich hunt". My personal experience tells me that extreme points of view are rarely valid.
>Now the downloadFailed is executed sequentially and queued on event loop, if you have some concerns please speak out.

DownloadHandler queues to self-destruct using the event loop after emiting this signal. Each outer DownloadHandler queues both self-destruct and re-emiting signal when it receives this signal. Are you sure this is all going to be done in the expected order, so some intermediate DownloadHandler won't be destroyed before it re-emits a signal?
I just want to make sure you understand exactly what you're doing, or at least that you've tested it with real multilevel redirection.

Reply:
>My personal experience tells me that extreme points of view are rarely valid.

No one ever said my viewpoint is the law, you're imagining things in your head.
It is from my experience that things are *usually* easier to maintain when you apply them and there are definitely exceptions although for me, not so common.

>Are you sure this is all going to be done in the expected order, so some intermediate DownloadHandler won't be destroyed before it re-emits a signal?

Thanks a lot!
I've tested that it works, but probably didn't cover all exec paths, now it does looks strange... I think I'll drop this change and worry later if it still pose a problem.